README.rst
==============================
Python time expressions parser
==============================


.. image:: https://img.shields.io/pypi/v/pytimeparser.svg
        :target: https://pypi.python.org/pypi/pytimeparser

.. image:: https://img.shields.io/travis/ouahibelhanchi/pytimeparser.svg
        :target: https://travis-ci.org/ouahibelhanchi/pytimeparser

.. image:: https://readthedocs.org/projects/pytimeparser/badge/?version=latest
        :target: https://pytimeparser.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status




Basic Python module to parse time expressions.


* Free software: MIT license
* Documentation: https://pytimeparser.readthedocs.io.


Usage
-----

To use Python time expressions parser in a project::

    import pytimeparser
    
    output_timedelta = pytimeparser.parse('3 days 5:12:43.123')

    print(output_timedelta.total_seconds())


    
Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.2.0 (2018-06-11)
------------------

* Raise `ValueError` in case no time expression recognized.

0.1.0 (2018-04-02)
------------------

* First release on PyPI.


tests/test_api.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `pytimeparser.api` module"""

import datetime

import pytest

from pytimeparser import parse


def test_parse_empty():
    with pytest.raises(ValueError):
        parse('')


def test_parse_non_string():
    with pytest.raises(TypeError):
        parse(None)


def test_parse_no_expression():
    with pytest.raises(ValueError):
        parse('no time')


def test_parse():
    output = parse('4 years, 3 months and 4 days')
    expected_output = datetime.timedelta(
        weeks=4*52+3*4, days=4,
    )
    assert output == expected_output


def test_parse_negative():
    output = parse('- 3 hours and 30 minutes and 300 microseconds')
    expected_output = datetime.timedelta(
        hours=-3, minutes=30, milliseconds=300*0.001
    )
    assert output == expected_output


def test_parse_whitespace():
    output = parse('-4    years  ,    \t 3 days   ')
    expected_output = datetime.timedelta(weeks=-4*52, days=3)
    assert output == expected_output


def test_parse_from_to():
    output = parse('3 to 4 days, 5 to 6 hours')
    expected_min = datetime.timedelta(days=3, hours=5)
    expected_max = datetime.timedelta(days=4, hours=6)
    assert expected_min <= output <= expected_max


def test_parse_free_text():
    output = parse('She said she will comback in 2 hours and 30 minutes')
    expected_output = datetime.timedelta(hours=2, minutes=30)
    assert output == expected_output


def test_parse_packed():
    output = parse('1w3d2h32m')
    expected_output = datetime.timedelta(weeks=1, days=3, hours=2, minutes=32)
    assert output == expected_output


def test_parse_packed_spaced():
    output = parse('1 w 3 d 2 h 32 m')
    expected_output = datetime.timedelta(weeks=1, days=3, hours=2, minutes=32)
    assert output == expected_output


def test_parse_duration():
    output = parse('2:04:13:2.266')
    expected_output = datetime.timedelta(
        hours=2, minutes=4, seconds=13, milliseconds=2, microseconds=266
    )
    assert output == expected_output


def test_parse_uptime_format():
    output = parse('2 days, 4:13:02')
    expected_output = datetime.timedelta(
        days=2, hours=4, minutes=13, seconds=2
    )
    assert output == expected_output


def test_parse_postgres_day_time_format():
    output = parse('3 days 04:05:06')
    expected_output = datetime.timedelta(
        days=3, hours=4, minutes=5, seconds=6
    )
    assert output == expected_output


def test_parse_iso_8601_duration_format():
    output = parse('P1Y2M3W10DT2H30M')
    expected_output = datetime.timedelta(
        weeks=1*52+2*4+3, days=10, hours=2, minutes=30
    )
    assert output == expected_output


tests/test_utils.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `pytimeparser.utils` module."""


from pytimeparser import utils


def test_remove_whitespace():
    output = utils.remove_whitespace('  this \t\r  is\fa\ntext  ')
    expected_output = 'thisisatext'
    assert output == expected_output


def test_normalize_whitespace():
    output = utils.normalize_whitespace('this    is  \n\f\t\r   a test')
    expected_output = 'this is a test'
    assert output == expected_output


def test_normalize_numbers():
    output = utils.normalize_numbers('text -\t112\f323 . 456 another +32 . 03')
    expected_output = 'text -112323.456 another +32.03'
    assert output == expected_output


def test_mapping_inversed():
    output = utils.mapping_inversed(dict(a=['aa', 'aaa'], b=['bb']))
    expected_output = {'aaa': 'a', 'aa': 'a', 'bb': 'b'}
    assert output == expected_output


def test_mapping_to_list():
    output = utils.mapping_to_list(dict(a='aa', b='bb', c='cc'))
    expected_output = ['a', 'b', 'c', 'aa', 'bb', 'cc']
    assert output.sort() == expected_output.sort()


pytimeparser/api.py
# -*- coding: utf-8 -*-


import re
import random
import datetime

from .utils import (
    normalize_text, expand_regex,
    mapping_inversed, mapping_to_list,
)


TIME_UNIT_MAPPING = dict(
    years=['y', 'year', 'years'],
    months=['month', 'months'],
    weeks=['w', 'wk', 'wks', 'week', 'weeks'],
    days=['d', 'day', 'days'],
    hours=['h', 'hr', 'hrs', 'hour', 'hours'],
    minutes=['m', 'min', 'mins', 'minute', 'minutes'],
    seconds=['s', 'sec', 'secs', 'second', 'seconds'],
    milliseconds=['ms', 'milli', 'millis', 'millisecond', 'milliseconds'],
    microseconds=['microsecond', 'microseconds'],
)

TIME_UNIT_INVERSED_MAPPING = mapping_inversed(TIME_UNIT_MAPPING)

TIME_UNIT_VARIATIONS = mapping_to_list(TIME_UNIT_INVERSED_MAPPING)

REGEX_SUPPORTED_TIME_UNIT_VARIATIONS = re.compile(
    r'|'.join([
        r'\b{}\b'.format(each)
        for each in TIME_UNIT_VARIATIONS
    ]),
    flags=re.UNICODE | re.IGNORECASE
)

REGEX_TIME_UNIT = re.compile(
    r'((?P<value_from>-?\d+(\.\d+)?)'
    r'(\s*to\s*(?P<value_to>-?\d+(\.\d+)?))?\s*)?'
    r'(?P<unit>{})'
    .format(REGEX_SUPPORTED_TIME_UNIT_VARIATIONS.pattern),
    flags=re.UNICODE | re.IGNORECASE
)

REGEX_DURATION = re.compile(
    r'((?P<hours>-?\d+(\.\d+)?)\s*:\s*(?P<minutes>-?\d+(\.\d+)?)'
    r'(\s*:\s*(?P<seconds>-?\d+(\.\d+)?)'
    r'(\s*:\s*(?P<milliseconds>-?\d+(\.\d+)?)'
    r'(\s*\.\s*(?P<microseconds>-?\d+(\.\d+)?))?)?)?)',
    flags=re.UNICODE | re.IGNORECASE
)

REGEX_ISO8601_DURATION = re.compile(
    r'(\s*P\s*'
    r'(\s*(?P<years>-?\d+(\.\d+)?)\s*Y\s*)?'
    r'(\s*(?P<months>-?\d+(\.\d+)?)\s*M\s*)?'
    r'(\s*(?P<weeks>-?\d+(\.\d+)?)\s*W\s*)?'
    r'(\s*(?P<days>-?\d+(\.\d+)?)\s*D\s*)?'
    r'(\s*T\s*'
    r'(\s*(?P<hours>-?\d+(\.\d+)?)\s*H\s*)?'
    r'(\s*(?P<minutes>-?\d+(\.\d+)?)\s*M\s*)?'
    r'(\s*(?P<seconds>-?\d+(\.\d+)?)\s*S\s*)?'
    r')?'
    r')',
    flags=re.UNICODE | re.IGNORECASE
)


def parse(text):
    """Parse time expression.

    :param text: Time expression to parse
    :return: :class:`datetime.timedelta` object
    :rtype: datetime.timedelta
    :raises: TypeError: if text is not a string
    :raises: ValueError: if text is empty
    """
    if not isinstance(text, ("".__class__, u"".__class__)):
        raise TypeError('Expected a string, received: %s' % type(text).__name__)

    text = normalize_text(text)
    if not text:
        raise ValueError('Expected a time expression, recieved an empty string')

    duration_regexs = (REGEX_DURATION, REGEX_ISO8601_DURATION,)
    for duration_regex in duration_regexs:
        text = expand_regex(duration_regex, text, '{value} {name}')

    kwargs = {}
    for match in REGEX_TIME_UNIT.finditer(text):
        unit = match.group('unit').lower()

        value_from = float(match.group('value_from') or '1')
        if match.group('value_to'):
            value_to = float(match.group('value_to'))
            value = random.uniform(value_from, value_to)
        else:
            value = value_from

        unit = TIME_UNIT_INVERSED_MAPPING[unit].lower()
        if unit == 'years':
            unit, value = 'weeks', value * 52
        elif unit == 'months':
            unit, value = 'weeks', value * 4
        elif unit == 'microseconds':
            unit, value = 'milliseconds', value * 0.001

        kwargs[unit] = kwargs.get(unit, 0) + value

    return datetime.timedelta(**kwargs)
